pkg/ingester/kafka_consumer.go

```diff
@@ -95,35 +97,89 @@ func (kc *kafkaConsumer) consume(ctx context.Context, records []partition.Record
 		minOffset    = int64(math.MaxInt64)
 		maxOffset    = int64(0)
 		consumeStart = time.Now()
+		limitWorkers = kc.maxConsumerWorkers
+		wg           sync.WaitGroup
 	)
 
+	// If the number of workers is set to 0, use the number of available CPUs
+	if limitWorkers == 0 {
+		limitWorkers = runtime.GOMAXPROCS(0)
+	}
+
+	// Find min/max offsets
 	for _, record := range records {
 		minOffset = min(minOffset, record.Offset)
 		maxOffset = max(maxOffset, record.Offset)
 	}
 
 	level.Debug(kc.logger).Log("msg", "consuming records", "min_offset", minOffset, "max_offset", maxOffset)
-	for _, record := range records {
-		stream, err := kc.decoder.DecodeWithoutLabels(record.Content)
-		if err != nil {
-			level.Error(kc.logger).Log("msg", "failed to decode record", "error", err)
-			continue
-		}
-		recordCtx := user.InjectOrgID(record.Ctx, record.TenantID)
-		req := &logproto.PushRequest{
-			Streams: []logproto.Stream{stream},
-		}
-		if err := retryWithBackoff(ctx, func(attempts int) error {
-			if _, err := kc.pusher.Push(recordCtx, req); err != nil {
-				level.Warn(kc.logger).Log("msg", "failed to push records", "err", err, "offset", record.Offset, "attempts", attempts)
-				return err
+
+	type recordWithIndex struct {
+		record partition.Record
+		index  int
+	}
+
+	numWorkers := min(limitWorkers, len(records))
+	workChan := make(chan recordWithIndex, numWorkers)
+	// success keeps track of the records that were processed. It is expected to
+	// be sorted in ascending order of offset since the records themselves are
+	// ordered.
+	success := make([]int64, len(records))
```

Comment:
I think one thing we just need to be careful of, and I didn't consider this at the start, is that we don't set a limit on the maximum number of records that we can poll from Kafka. I would suggest testing this in dev (and also putting a custom image in ops with `flux-ignore`) to test this.

Reply:
Yes, I was planning to test on dev already before we move forward.
I will consider take a look on ops too, thanks